instrumentation-api-annotation-support/src/main/java/io/opentelemetry/instrumentation/api/annotation/support/MethodSpanAttributesExtractor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.annotation.support;
+
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.instrumentation.api.caching.Cache;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/** Extractor of {@link io.opentelemetry.api.common.Attributes} for a traced method. */
+public class MethodSpanAttributesExtractor<REQUEST, RESPONSE>
+    extends AttributesExtractor<REQUEST, RESPONSE> {
+
+  private final BaseAttributeBinder binder;
+  private final MethodExtractor<REQUEST> methodExtractor;
+  private final MethodArgumentsExtractor<REQUEST> methodArgumentsExtractor;
+  private final Cache<Method, AttributeBindings> cache;
+
+  /** Returns a new {@link MethodSpanAttributesExtractorBuilder}. */
+  public static <REQUEST, RESPONSE>
+      MethodSpanAttributesExtractorBuilder<REQUEST, RESPONSE> newBuilder(
+          MethodExtractor<REQUEST> methodResolver) {
+
+    return new MethodSpanAttributesExtractorBuilder<>(methodResolver);
+  }
+
+  MethodSpanAttributesExtractor(MethodSpanAttributesExtractorBuilder<REQUEST, RESPONSE> builder) {
+    this.methodExtractor = builder.methodExtractor;
+    this.methodArgumentsExtractor = builder.methodArgumentsExtractor;
+    this.binder = new MethodSpanAttributeBinder(builder.parameterAttributeNamesExtractor);
+    this.cache = Cache.newBuilder().setWeakKeys().build();
+  }
+
+  @Override
+  protected void onStart(AttributesBuilder attributes, REQUEST request) {
+    Method method = methodExtractor.extract(request);
+    AttributeBindings bindings;
+    if (cache != null) {
+      bindings = cache.computeIfAbsent(method, binder::bind);
```

Comment:
Our weak caches use identity comparison for keys. For this cache to work it would need to get the exact same method instance. For example `Class.getDeclaredMethod` returns a different instance each time. Are you sure that  `MethodExtractor` implementations will actually always return the same instance? Should this be documented? Or use a different caching strategy, for example could use a `ClassValue` on declaring class of the method that contains a plain concurrent map. If you believe that the cache should be weak then you could use `ClassValue` to hold keys that you derive from method so that different instances of the same method would have the same key.

Reply:
Oof, I did not know that our cache implementation was limited to identity comparison of weak keys.  I see that this behavior is effectively inherited from Caffeine so we don't have an easy way to support value-based equality of weak keys.

I ran some tests locally and it seems like Spring AOP does cache the Method instance on subsequent invocations, but I am a little hesitant to depend on this behavior.

Reply:
Unfortunately ByteBuddy does not cache the Method instance.  When referencing `@Advice.Origin Method method` it effectively emits `Foo.class.getMethod("name", ...)` on each invocation.  I'll look into your suggestions on different cache strategies.

Reply:
I added an implementation of `Cache<K, V>` that uses `ClassValue<T>` under the hood.  It's probably overkill to have it implement `Cache<K, V>` since I'll only ever use `computeIfAbsent` here.